File: Section4/src/main/java/proxy/Proxy.java
package proxy;

public class Proxy implements Subject{
    private Subject subject;
    Proxy() {
        subject = new RealSubject();
    }
    @Override
    public void request() {
        doBefore();
        subject.request();
        doNext();
    }

    private void doBefore() {
    }

    private void doNext() {
    }
}

File: Kotlin/src/structural/proxy/Proxy.kt
package structural.proxy

class Proxy(realSubject: RealSubject) : Subject {
    private var realSubject : Subject? = realSubject

    private fun doFirst() {
        println("Proxy do first ...")
    }

    override fun request() {
        doFirst()
        realSubject!!.request()
        doLast()
    }

    private fun doLast() {
        println("Proxy do last ...")
    }

}